### PR TITLE
Fixed #218 -- Add custom label support

### LINKED ISSUE #218: Add custom label support
Based on #208, waiting for supporters

### PR DESCRIPTION
**Pull request by mostafa-anm**

### CODE REVIEW

**Comment by codingjoe** [comment]
@mostafa-anm please add a very detailed docstring, including an example on how to overwrite the behavior and a test of course.

**Comment by mostafa-anm** [comment]
@codingjoe added...

**Comment by codingjoe** [comment]
You don't use it everywhere make sure to extend you tests to cover the spots you missed:
- https://github.com/applegrew/django-select2/blob/master/django_select2/forms.py#L234
- https://github.com/applegrew/django-select2/blob/master/django_select2/forms.py#L392
- https://github.com/applegrew/django-select2/blob/master/django_select2/forms.py#L397

File: django_select2/forms.py [hunk 1]
```diff
@@ -399,6 +399,20 @@ def render_options(self, choices, selected_choices):
...
+
+        Can be overriden to change the representation of each choice.
+
+        Example usage::
+
+            class MyWidget(ModelSelect2Widget):
+                def label_from_instance(obj):
+                    return smart_text(obj.title.upper())
+        """
+        return smart_text(obj)
```
Comment (codingjoe) [inline django_select2/forms.py]: please use `force_text`

File: django_select2/forms.py [hunk 2]
```diff
@@ -399,6 +399,20 @@ def render_options(self, choices, selected_choices):
...
+        """
+        Label representation from instance.
+
+        Can be overriden to change the representation of each choice.
+
+        Example usage::
+
+            class MyWidget(ModelSelect2Widget):
+                def label_from_instance(obj):
+                    return smart_text(obj.title.upper())
```
Comment (codingjoe) [inline django_select2/forms.py]: please a trailing newline for code black

**Comment by codingjoe** [comment]
@mostafa-anm is this done now? Should I review it again?

**Comment by mostafa-anm** [comment]
@codingjoe yes, it now label_from_instance covers also, ModelSelect2Mixin and HeavySelect2Widget.

File: django_select2/forms.py [hunk 3]
```diff
@@ -108,6 +108,21 @@ def _get_media(self):
 
     media = property(_get_media)
 
+    def label_from_instance(self, obj):
+        """
+        Label representation from instance.
```
Comment (codingjoe) [inline django_select2/forms.py]: `Return label representation from given instance.`

File: django_select2/forms.py [hunk 4]
```diff
@@ -108,6 +108,21 @@ def _get_media(self):
...
+
+        Can be overridden to change the representation of each choice.
```
Comment (codingjoe) [inline django_select2/forms.py]: Don't forget to describe behavior of the default implementation. Make sure users get, that instances need have a string representation (to implement `__str__`).

File: django_select2/forms.py [hunk 5]
```diff
@@ -108,6 +108,21 @@ def _get_media(self):
 
     media = property(_get_media)
 
+    def label_from_instance(self, obj):
+        """
+        Label representation from instance.
+
```
Comment (codingjoe) [inline django_select2/forms.py]: Make sure to drop a line that this will be used to create the select options. Focus on these to words, as they are what people will be looking fore, since they come from an html world and know nothing about this code.

File: django_select2/forms.py [hunk 6]
```diff
@@ -228,6 +243,10 @@ def set_to_cache(self):
 
     def render_options(self, choices, selected_choices):
         """Render only selected options."""
+        if choices:
+            choices = [(k, self.label_from_instance(v)) for k, v in choices]
+        if self.choices:
```
Comment (codingjoe) [inline django_select2/forms.py]: why is this needed? This will result in a massive query. The whole result set will be evaluated. No can do, sorry ;)

Comment (codingjoe) [inline django_select2/forms.py]: why is this needed? This will result in a massive query. The whole result set will be evaluated. No can do, sorry ;)

File: django_select2/forms.py (hunk 5 above)
Comment (codingjoe) [inline django_select2/forms.py]: Make sure to drop a line that this will be used to create the select options. Focus on these to words, as they are what people will be looking fore, since they come from an html world and know nothing about this code.

File: tests/test_forms.py [hunk 7]
```diff
@@ -104,6 +108,12 @@ def test_selected_option(self, db):
                not_required_field.widget.render('primary_genre', 1, choices=NUMBER_CHOICES), \
             not_required_field.widget.render('primary_genre', 1, choices=NUMBER_CHOICES)
 
+    def test_selected_option_label_from_instance(self, db):
```
Comment (codingjoe) [inline tests/test_forms.py]: That is already tested. No need.

File: tests/test_forms.py [hunk 8]
```diff
@@ -91,6 +91,10 @@ class TestHeavySelect2Mixin(TestSelect2Mixin):
     def test_initial_data(self):
         assert 'One' in self.form.as_p()
 
+    def test_initial_data_label_from_instance(self):
+        form = forms.HeavySelect2WidgetForm(initial={'artist': 1})
```
Comment (codingjoe) [inline tests/test_forms.py]: This means the test for the 'light' widget is missing.

File: django_select2/forms.py [hunk 9]
```diff
@@ -108,6 +108,21 @@ def _get_media(self):
 
     media = property(_get_media)
 
+    def label_from_instance(self, obj):
```
Comment (codingjoe) [inline django_select2/forms.py]: I'm missing the implementation for the `light` version. This should be really consistent.

**Comment by codingjoe** [comment]
@mostafa-anm good work. This is very close. I'm just missing the implementation for the "light" widgets. This isn't a big deal, just check out how the options are being rendered for the `Select2WidgetMixin`.
A lot of people would like this feature, I'd really like to finish this!

**Comment by codingjoe** [comment]
I changed it to only support the model field as this is the only one that handles model instance anyways.

Released in 5.6.0
